### PR TITLE
Tighten package sources in C# smoketests

### PR DESCRIPTION
**Pull request by RReverser**
# Description of Changes

Before this, we couple of times ran into issues where a version in SpacetimeDB template was outdated, or NuGet sources were in incorrect order, and so on, which led to either tests failing with misleading error (https://github.com/clockworklabs/SpacetimeDB/commit/2b09c8d1f8868c22f8d3b6af30a492ef1ac9569c being most recent example), or, worse, quietly passing even when code or versions are actually broken.

Apparently, NuGet has a package source mapping feature which allows to ensure that specific packages are only pulled from specific custom sources, which is exactly what we want here. It allows us to ensure that SpacetimeDB.* pacakges for smoketests are pulled from our local build folders, and not from nuget.org.

# API and ABI breaking changes

If this is an API or ABI breaking change, please apply the
corresponding GitHub label.

# Expected complexity level and risk

*How complicated do you think these changes are? Grade on a scale from 1 to 5,
where 1 is a trivial change, and 5 is a deep-reaching and complex change.*

*This complexity rating applies not only to the complexity apparent in the diff,
but also to its interactions with existing and future code.*

*If you answered more than a 2, explain what is complex about the PR,
and what other components it interacts with in potentially concerning ways.*

# Testing

*Describe any testing you've done, and any testing you'd like your reviewers to do,
so that you're confident that all the changes work as expected!*

- [x] *Write a test you've completed here.*
- [ ] *Write a test you want a reviewer to do here, so they can check it off when they're satisfied.*

### CODE REVIEW

**Review by mamcx — APPROVED** [review]
LGTM

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -137,9 +137,14 @@ jobs:
...
-              <!-- Official NuGet.org server -->
-              <add key="NuGet.org" value="https://api.nuget.org/v3/index.json" />
             </packageSources>
+            <packageSourceMapping>
+              <!-- Ensure that SpacetimeDB.BSATN.Runtime is used from the local folder. -->
+              <!-- Otherwise we risk an outdated version being quietly pulled from NuGet for testing. -->
+              <packageSource key="Local SpacetimeDB.BSATN.Runtime">
+                <package pattern="SpacetimeDB.BSATN.Runtime" />
+              </packageSource>
+            </packageSourceMapping>
```
Comment (bfops) [inline .github/workflows/ci.yml:147]: FWIW, the previous nuget-fallback behavior was entirely intentional (see the PR description that introduced it: https://github.com/clockworklabs/SpacetimeDB/pull/1503).

tl;dr if we release breaking changes in this repo, the SDK tests would continue to use a package they were compatible with, rather than breaking for every future SpacetimeDB PR until someone fixed the C# SDK.

Comment (bfops) [reply .github/workflows/ci.yml:147]: This way, the C# SDK tests _only_ break if you fail to bump the version numbers when you make a breaking change.

I don't feel strongly about keeping it that way (and I believe it's not a required check, so this doesn't "deadlock" us), but that was the original intent.

Comment (RReverser) [reply .github/workflows/ci.yml:147]: > This way, the C# SDK tests only break if you fail to bump the version numbers when you make a breaking change.

Problem is, we also have the opposite sometimes - we bumped SpacetimeDB version, made some breaking changes, but C# SDK tests are still passing "successfully" because they are still pointing to old BSATN.* packages.

This is too easy to miss at the moment, so I do think the noisy error would be better in this context.

Comment (RReverser) [reply .github/workflows/ci.yml:147]: Although I can see the arguments for the original behaviour too...

Comment (bfops) [reply .github/workflows/ci.yml:147]: Sounds good - we should probably add a requirement that people breaking a downstream repo (in this case, the C# SDK) have a fix PR ready to go before merging their breaking change in this repo.

Comment (RReverser) [reply .github/workflows/ci.yml:147]: I'd say let's try this stricter mode to catch breaking behaviour sooner; if it does prove too annoying, we can always revert those changes.
